File: app/models/embargos_execucao.py
"""Fluxo Embargos à Execução (emb_*).

Um card por PASTA de execução do BB Autor cuja tarefa "Protocolar Inicial -
BB Autor" foi cumprida no L1 (= ajuizamento). A partir da data do ajuizamento
o Flow espera N dias úteis e passa a consultar o tribunal (DataJud + DJEN) a
cada 5 dias úteis até achar os embargos à execução ligados a ela. Ver
docs/embargos-execucao-plano.md.
"""

from sqlalchemy import (
    BigInteger,
    Boolean,
    Column,
    Date,
    DateTime,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
    func,
    true,
)
from sqlalchemy.orm import relationship

from app.db.session import Base
from app.db.types import jsonb

# ── De onde o card veio ──────────────────────────────────────────────
ORIGEM_RELATORIO = "RELATORIO"   # relatório "ROBÔ - EMBARGOS À EXECUÇÃO" do L1
ORIGEM_PLANILHA = "PLANILHA"     # legado subido pelo operador
ORIGEM_MANUAL = "MANUAL"

# ── Estado do monitoramento ──────────────────────────────────────────
ESTADO_SEM_CNJ = "SEM_CNJ"                    # sem número: não dá pra consultar o tribunal
ESTADO_AGUARDANDO_JANELA = "AGUARDANDO_JANELA"  # antes dos N dias úteis
ESTADO_MONITORANDO = "MONITORANDO"            # consulta a cada X dias úteis
ESTADO_ENCONTRADO = "ENCONTRADO"              # candidato forte — parou e avisou
ESTADO_CONFIRMADO = "CONFIRMADO"              # operador validou o vínculo
ESTADO_CONCLUIDO = "CONCLUIDO"                # tarefas disparadas no incidente — fluxo fechado
ESTADO_JA_CADASTRADO = "JA_CADASTRADO"        # a pasta já tem incidente de embargos no L1
ESTADO_SEM_EMBARGOS = "SEM_EMBARGOS"          # teto do monitoramento sem achar nada
ESTADO_ENCERRADO = "ENCERRADO"                # tirado do fluxo pelo operador

ESTADOS = (
    ESTADO_SEM_CNJ,
    ESTADO_AGUARDANDO_JANELA,
    ESTADO_MONITORANDO,
    ESTADO_ENCONTRADO,
    ESTADO_CONFIRMADO,
    ESTADO_CONCLUIDO,
    ESTADO_JA_CADASTRADO,
    ESTADO_SEM_EMBARGOS,
    ESTADO_ENCERRADO,
)

# ── Partes demandadas no portal do BB ────────────────────────────────
PARTES_PENDENTE = "PENDENTE"
PARTES_OK = "OK"
PARTES_ERRO = "ERRO"            # a coleta falhou — volta na próxima passagem (com teto)
PARTES_SEM_NPJ = "SEM_NPJ"      # pasta sem NPJ: não há o que consultar no portal
PARTES_NAO_APLICA = "NAO_APLICA"  # cliente não é o BB (Banese/Ativos)

# ── Força da evidência de um candidato ───────────────────────────────
NIVEL_CONFIRMADO_DJEN = "CONFIRMADO_DJEN"  # embargante do DJEN é parte da execução
NIVEL_PROVAVEL = "PROVAVEL"                # dependência + petição na execução no mesmo dia
NIVEL_FRACO = "FRACO"                      # só vara + classe + data
NIVEL_DESCARTADO = "DESCARTADO"            # DJEN mostra embargante que não é da execução
# Os embargos chegaram pela fila de Publicações (controle unificado) — vira
# candidato da execução pra encerrar o monitor sem repetir o aviso.
NIVEL_PUBLICACAO = "PUBLICACAO"
NIVEIS_FORTES = (NIVEL_CONFIRMADO_DJEN, NIVEL_PROVAVEL, NIVEL_PUBLICACAO)

DECISAO_PENDENTE = "PENDENTE"
DECISAO_CONFIRMADO = "CONFIRMADO"
DECISAO_RECUSADO = "RECUSADO"

# ── Trilha de eventos ────────────────────────────────────────────────
SECAO_ENTRADA = "ENTRADA"
SECAO_PARTES = "PARTES"
SECAO_TRIBUNAL = "TRIBUNAL"
SECAO_L1 = "L1"
SECAO_AVISO = "AVISO"
SECAO_TAREFAS = "TAREFAS"
SECAO_OPERADOR = "OPERADOR"
SECAO_CONTROLE = "CONTROLE"

# ── Controle de Embargos (visão única da Controladoria) ──────────────
# Decisões do operador (14/09/2026): um caso por execução + embargos; se a
# pasta incidental (ou a pasta dos embargos) existe no L1, o trabalho da
# Controladoria está feito, ponto; publicação caída na pasta da execução sem
# incidente é FALHA DE CADASTRO; só BB Autor por enquanto; o mesmo embargo
# vindo por duas filas vira um caso só (a 1ª fonte cria a tarefa, a 2ª vincula).
CASO_PENDENTE = "PENDENTE"        # embargos sem pasta incidental no L1
CASO_CADASTRADO = "CADASTRADO"    # a pasta existe — fim
CASO_DESCARTADO = "DESCARTADO"    # não eram embargos desta carteira

ORIG_TRIBUNAL = "TRIBUNAL"            # monitor do tribunal (DataJud + DJEN)
ORIG_PUB_SEM_PASTA = "PUB_SEM_PASTA"  # publicação sem pasta, tipo Embargos à Execução
ORIG_PUB_NA_PASTA = "PUB_NA_PASTA"    # publicação caída na pasta da EXECUÇÃO (falha de cadastro)
ORIG_PUB_INCIDENTE = "PUB_INCIDENTE"  # publicação já na pasta dos embargos
ORIG_L1 = "L1"                        # incidente achado direto no Legal One
ORIGENS_COLUNA = {
    ORIG_TRIBUNAL: "origem_tribunal",
    ORIG_PUB_SEM_PASTA: "origem_pub_sem_pasta",
    ORIG_PUB_NA_PASTA: "origem_pub_na_pasta",
    ORIG_PUB_INCIDENTE: "origem_pub_incidente",
    ORIG_L1: "origem_l1",
}
ORIGENS_PUBLICACAO = (ORIG_PUB_SEM_PASTA, ORIG_PUB_NA_PASTA, ORIG_PUB_INCIDENTE)
PUB_IGNORADA_MONITORIA = "IGNORADA_MONITORIA"  # embargos à monitória: fora do controle
# Medido no Docker local em 14/09/2026 com as 95 publicações reais do BB Autor:
# a subcategoria de Publicações mistura embargos de declaração, e parte dos
# embargos à execução foi oposta NOS PRÓPRIOS AUTOS (sem processo apartado,
# logo sem pasta incidental a cadastrar).
PUB_IGNORADA_NAO_EMBARGOS = "IGNORADA_NAO_EMBARGOS_EXECUCAO"
PUB_IGNORADA_PROPRIOS_AUTOS = "IGNORADA_PROPRIOS_AUTOS"

EVT_INFO = "INFO"
EVT_AVISO = "AVISO"
EVT_ERRO = "ERRO"

# ── Tarefas do incidente (templates) ─────────────────────────────────
RESP_FIXO = "FIXO"                          # pessoa escolhida no template
RESP_ADVOGADO_CARD = "ADVOGADO_RESPONSAVEL"  # advogado responsável da execução (relatório)
DISPARO_CRIADA = "CRIADA"
DISPARO_FALHA = "FALHA"


class EmbExecucao(Base):
    __tablename__ = "emb_execucao"

    id = Column(Integer, primary_key=True)
    # Chave natural: a pasta ("Proc - 0068694"). Sem pasta (planilha só com
    # CNJ) vira "CNJ <número>" e é adotada quando a pasta aparecer.
    pasta = Column(String, nullable=False, unique=True, index=True)
    l1_task_id = Column(BigInteger, nullable=True, index=True)
    l1_task_ids = Column(jsonb(), nullable=True)
    lawsuit_id = Column(Integer, nullable=True)
    cnj = Column(String, nullable=True)
    cnj_digitos = Column(String, nullable=True, index=True)
    npj = Column(String, nullable=True, index=True)
    escritorio = Column(String, nullable=True)
    cliente = Column(String, nullable=True, index=True)
    responsavel_nome = Column(String, nullable=True)
    executante_nome = Column(String, nullable=True)
    uf = Column(String, nullable=True)
    origem = Column(String, nullable=False, server_default=ORIGEM_RELATORIO)

    data_ajuizamento = Column(Date, nullable=False)
    estado = Column(String, nullable=False, server_default=ESTADO_AGUARDANDO_JANELA, index=True)
    dias_uteis_janela = Column(Integer, nullable=False, server_default="15")
    inicio_monitoramento = Column(Date, nullable=True)
    proxima_consulta = Column(Date, nullable=True, index=True)
    ultima_consulta_em = Column(DateTime(timezone=True), nullable=True)
    consultas_feitas = Column(Integer, nullable=False, server_default="0")
    ultimo_erro = Column(Text, nullable=True)

    partes_status = Column(String, nullable=False, server_default=PARTES_PENDENTE, index=True)
    partes_tentativas = Column(Integer, nullable=False, server_default="0")
    partes_erro = Column(Text, nullable=True)
    partes_em = Column(DateTime(timezone=True), nullable=True)

    # Capa da execução no DataJud (guardada pra busca de candidatos por vara).
    tribunal_alias = Column(String, nullable=True)
    orgao_codigo = Column(Integer, nullable=True)
    orgao_nome = Column(String, nullable=True)
    datajud_ajuizamento_raw = Column(String, nullable=True)

    # Incidente de embargos já existente na pasta (ProceduralIssues do L1).
    l1_incidente_verificado_em = Column(DateTime(timezone=True), nullable=True)
    incidente_folder = Column(String, nullable=True)
    incidente_cnj = Column(String, nullable=True)
    incidente_id = Column(Integer, nullable=True)
    incidente_office_id = Column(Integer, nullable=True)

    encontrado_em = Column(DateTime(timezone=True), nullable=True)
    aviso_enviado_em = Column(DateTime(timezone=True), nullable=True)
    confirmado_candidato_id = Column(Integer, nullable=True)

    anotacao = Column(Text, nullable=True)
    decidido_por_user_id = Column(Integer, ForeignKey("legal_one_users.id"), nullable=True)
    decidido_em = Column(DateTime(timezone=True), nullable=True)

    criado_em = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    atualizado_em = Column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    partes = relationship(
        "EmbParte", back_populates="execucao", cascade="all, delete-orphan",
        order_by="EmbParte.id",
    )
    candidatos = relationship(
        "EmbCandidato", back_populates="execucao", cascade="all, delete-orphan",
        order_by="EmbCandidato.id",
    )


class EmbParte(Base):
    __tablename__ = "emb_parte"

    id = Column(Integer, primary_key=True)
    execucao_id = Column(
        Integer, ForeignKey("emb_execucao.id", ondelete="CASCADE"), nullable=False, index=True
    )
    origem = Column(String, nullable=False, server_default="BB")  # BB | DJEN
    polo = Column(String, nullable=True)
    nome = Column(String, nullable=False)
    cpf_cnpj = Column(String, nullable=True)
    tipo_pessoa = Column(String, nullable=True)
    relacao_bb = Column(String, nullable=True)
    demandada = Column(Boolean, nullable=False, server_default="0")
    raw = Column(jsonb(), nullable=True)
    criado_em = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    execucao = relationship("EmbExecucao", back_populates="partes")


class EmbCandidato(Base):
    __tablename__ = "emb_candidato"
    __table_args__ = (
        UniqueConstraint("execucao_id", "cnj_digitos", name="uq_emb_candidato_execucao_cnj"),
    )

    id = Column(Integer, primary_key=True)
    execucao_id = Column(
        Integer, ForeignKey("emb_execucao.id", ondelete="CASCADE"), nullable=False, index=True
    )
    cnj = Column(String, nullable=False)
    cnj_digitos = Column(String, nullable=False, index=True)
    data_ajuizamento = Column(DateTime(timezone=True), nullable=True)
    classe_nome = Column(String, nullable=True)
    orgao_nome = Column(String, nullable=True)
    distribuicao_dependencia = Column(Boolean, nullable=False, server_default="0")
    peticao_mesmo_dia = Column(Boolean, nullable=False, server_default="0")
    nivel = Column(String, nullable=False, server_default=NIVEL_FRACO, index=True)

    djen_status = Column(String, nullable=True)
    djen_embargantes = Column(jsonb(), nullable=True)
    djen_embargados = Column(jsonb(), nullable=True)
    # Embargos já cadastrados no L1 (achados pelo CNJ assim que viram candidato forte).
    l1_litigation_id = Column(Integer, nullable=True)
    l1_folder = Column(String, nullable=True)
    djen_nomes_casados = Column(jsonb(), nullable=True)
    djen_trecho = Column(Text, nullable=True)
    djen_data = Column(String, nullable=True)
    djen_consultado_em = Column(DateTime(timezone=True), nullable=True)

    decisao = Column(String, nullable=False, server_default=DECISAO_PENDENTE, index=True)
    decidido_por_user_id = Column(Integer, ForeignKey("legal_one_users.id"), nullable=True)
    decidido_em = Column(DateTime(timezone=True), nullable=True)

    criado_em = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    atualizado_em = Column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    execucao = relationship("EmbExecucao", back_populates="candidatos")


class EmbEvento(Base):
    __tablename__ = "emb_evento"

    id = Column(Integer, primary_key=True)
    execucao_id = Column(
        Integer, ForeignKey("emb_execucao.id", ondelete="CASCADE"), nullable=True, index=True
    )
    caso_id = Column(
        Integer, ForeignKey("emb_caso.id", ondelete="SET NULL"), nullable=True, index=True
    )
    secao = Column(String, nullable=False, index=True)
    nivel = Column(String, nullable=False, server_default=EVT_INFO)
    mensagem = Column(Text, nullable=False)
    dados = Column(jsonb(), nullable=True)
    user_id = Column(Integer, ForeignKey("legal_one_users.id"), nullable=True)
    criado_em = Column(DateTime(timezone=True), server_default=func.now(), nullable=False, index=True)


class EmbTarefaTemplate(Base):
    """Tarefa que o Flow cria no incidente depois do cadastro (tela de templates)."""

    __tablename__ = "emb_tarefa_template"

    id = Column(Integer, primary_key=True)
    nome = Column(String, nullable=False)
    ativo = Column(Boolean, nullable=False, server_default=true())
    ordem = Column(Integer, nullable=False, server_default="0")
    tipo_id = Column(Integer, nullable=False)
    subtipo_id = Column(Integer, nullable=False)
    subtipo_nome = Column(String, nullable=True)
    responsavel_modo = Column(String, nullable=False, server_default=RESP_FIXO)
    responsavel_contact_id = Column(Integer, nullable=True)
    responsavel_nome = Column(String, nullable=True)
    prazo_dias_uteis = Column(Integer, nullable=False, server_default="5")
    prioridade = Column(String, nullable=False, server_default="Normal")
    descricao_template = Column(Text, nullable=False)
    observacoes_template = Column(Text, nullable=True)
    atualizado_por_user_id = Column(Integer, ForeignKey("legal_one_users.id"), nullable=True)
    criado_em = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    atualizado_em = Column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )


class EmbTarefaDisparo(Base):
    """Cada tarefa enviada (ou tentada) ao L1 — cópia do que saiu, sobrevive ao template."""

    __tablename__ = "emb_tarefa_disparo"

    id = Column(Integer, primary_key=True)
    execucao_id = Column(
        Integer, ForeignKey("emb_execucao.id", ondelete="CASCADE"), nullable=False, index=True
    )
    template_id = Column(Integer, nullable=True)
    template_nome = Column(String, nullable=True)
    incidente_id = Column(Integer, nullable=True)
    subtipo_id = Column(Integer, nullable=True)
    responsavel_contact_id = Column(Integer, nullable=True)
    prazo = Column(Date, nullable=True)
    descricao = Column(Text, nullable=True)
    status = Column(String, nullable=False, index=True)
    l1_task_id = Column(BigInteger, nullable=True)
    erro = Column(Text, nullable=True)
    user_id = Column(Integer, ForeignKey("legal_one_users.id"), nullable=True)
    criado_em = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)


class EmbCaso(Base):
    """Um caso de embargos à execução do BB Autor, venha de onde vier."""

    __tablename__ = "emb_caso"

    id = Column(Integer, primary_key=True)
    # CNJ dos embargos. Nulo só quando a publicação caiu na pasta da execução e
    # o texto não traz o número dos embargos (o caso fica pela pasta).
    cnj_embargos = Column(String, nullable=True)
    cnj_embargos_digitos = Column(String, nullable=True, unique=True, index=True)
    execucao_id = Column(
        Integer, ForeignKey("emb_execucao.id", ondelete="SET NULL"), nullable=True, index=True
    )
    pasta_execucao = Column(String, nullable=True, index=True)
    cnj_execucao = Column(String, nullable=True)
    lawsuit_id_execucao = Column(Integer, nullable=True, index=True)

    estado = Column(String, nullable=False, server_default=CASO_PENDENTE, index=True)
    falha_cadastro = Column(Boolean, nullable=False, server_default="0", index=True)
    vinculo_confirmado = Column(Boolean, nullable=False, server_default="0")
    origem_tribunal = Column(Boolean, nullable=False, server_default="0")
    origem_pub_sem_pasta = Column(Boolean, nullable=False, server_default="0")
    origem_pub_na_pasta = Column(Boolean, nullable=False, server_default="0")
    origem_pub_incidente = Column(Boolean, nullable=False, server_default="0")
    origem_l1 = Column(Boolean, nullable=False, server_default="0")
    primeira_origem = Column(String, nullable=True)

    embargante = Column(String, nullable=True)
    candidato_id = Column(Integer, nullable=True)
    tarefa_l1_id = Column(BigInteger, nullable=True)
    detectado_em = Column(DateTime(timezone=True), nullable=True)

    incidente_id = Column(Integer, nullable=True)
    incidente_folder = Column(String, nullable=True)
    cadastrado_em = Column(DateTime(timezone=True), nullable=True)
    verificado_l1_em = Column(DateTime(timezone=True), nullable=True)
    verificacoes_l1 = Column(Integer, nullable=False, server_default="0")
    vinculo_tentado_em = Column(DateTime(timezone=True), nullable=True)

    descartado_motivo = Column(Text, nullable=True)
    decidido_por_user_id = Column(Integer, ForeignKey("legal_one_users.id"), nullable=True)
    criado_em = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    atualizado_em = Column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    execucao = relationship("EmbExecucao")
    publicacoes = relationship(
        "EmbCasoPublicacao", back_populates="caso", cascade="all, delete-orphan",
        order_by="EmbCasoPublicacao.id",
    )


class EmbCasoPublicacao(Base):
    """Publicação já lida pelo controle (uma publicação entra uma vez só)."""

    __tablename__ = "emb_caso_publicacao"

    id = Column(Integer, primary_key=True)
    caso_id = Column(
        Integer, ForeignKey("emb_caso.id", ondelete="CASCADE"), nullable=True, index=True
    )
    publicacao_id = Column(Integer, nullable=False, unique=True, index=True)
    origem = Column(String, nullable=False)
    data_publicacao = Column(String, nullable=True)
    status_publicacao = Column(String, nullable=True)
    linked_lawsuit_id = Column(Integer, nullable=True)
    trecho = Column(Text, nullable=True)
    criado_em = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    caso = relationship("EmbCaso", back_populates="publicacoes")

File: frontend/src/services/embargos-execucao.ts
// Serviço do Fluxo Embargos à Execução (aba da Controladoria no Minha Equipe).
// Backend: /api/v1/embargos-execucao (gate do time bb-cadastro).

import { apiFetch } from "@/lib/api-client";

const BASE = "/api/v1/embargos-execucao";

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) {
    let detail = `Erro ${res.status}`;
    try {
      detail = (await res.json())?.detail || detail;
    } catch {
      /* ignore */
    }
    throw new Error(detail);
  }
  return res.json() as Promise<T>;
}

export const ESTADOS = [
  "ENCONTRADO",
  "MONITORANDO",
  "AGUARDANDO_JANELA",
  "SEM_CNJ",
  "CONFIRMADO",
  "CONCLUIDO",
  "JA_CADASTRADO",
  "SEM_EMBARGOS",
  "ENCERRADO",
] as const;

export const ESTADO_LABEL: Record<string, string> = {
  ENCONTRADO: "Embargos encontrados",
  MONITORANDO: "Monitorando",
  AGUARDANDO_JANELA: "Aguardando janela",
  SEM_CNJ: "Sem CNJ",
  CONFIRMADO: "Vínculo confirmado",
  CONCLUIDO: "Tarefas disparadas",
  JA_CADASTRADO: "Já cadastrado no L1",
  SEM_EMBARGOS: "Sem embargos (teto)",
  ENCERRADO: "Encerrado",
};

export const ESTADO_HINT: Record<string, string> = {
  ENCONTRADO: "O tribunal tem embargos com evidência forte de vínculo — o monitoramento parou e espera a conferência.",
  MONITORANDO: "A janela já passou: o Flow consulta o tribunal a cada intervalo de dias úteis.",
  AGUARDANDO_JANELA: "Ainda dentro dos 15/20/25 dias úteis após o ajuizamento.",
  SEM_CNJ: "A pasta entrou sem número de processo — o Flow procura o CNJ no Legal One uma vez por dia.",
  CONFIRMADO: "Vínculo confirmado: cadastre o incidente no Legal One e no portal do BB e dispare as tarefas.",
  CONCLUIDO: "Incidente cadastrado e tarefas disparadas — fluxo fechado.",
  JA_CADASTRADO: "A pasta já tem incidente de embargos cadastrado no Legal One.",
  SEM_EMBARGOS: "Chegou ao teto do monitoramento sem achar embargos.",
  ENCERRADO: "Tirado do fluxo pelo operador.",
};

export const PARTES_LABEL: Record<string, string> = {
  PENDENTE: "Aguardando portal BB",
  OK: "Partes lidas",
  ERRO: "Erro no portal BB",
  SEM_NPJ: "Sem NPJ",
  NAO_APLICA: "Não é BB",
};

export const NIVEL_LABEL: Record<string, string> = {
  CONFIRMADO_DJEN: "Embargante é parte (DJEN)",
  PROVAVEL: "Provável (dependência + petição no dia)",
  FRACO: "A verificar (só vara + classe + data)",
  DESCARTADO: "Descartado (embargante de outro processo)",
  DESCARTADO_EMBARGADO: "Descartado (embargado não é o cliente)",
};

export const PRIORIDADE_LABEL: Record<string, string> = { Low: "Baixa", Normal: "Normal", High: "Alta" };

export interface EmbargosItem {
  id: number;
  pasta: string;
  cnj: string | null;
  npj: string | null;
  cliente: string | null;
  escritorio: string | null;
  responsavel_nome: string | null;
  executante_nome: string | null;
  uf: string | null;
  origem: string;
  l1_task_id: number | null;
  data_ajuizamento: string;
  estado: string;
  dias_uteis_janela: number;
  inicio_monitoramento: string | null;
  proxima_consulta: string | null;
  ultima_consulta_em: string | null;
  consultas_feitas: number;
  ultimo_erro: string | null;
  partes_status: string;
  partes_tentativas: number;
  partes_erro: string | null;
  partes_em: string | null;
  tribunal: string | null;
  orgao_nome: string | null;
  encontrado_em: string | null;
  aviso_enviado_em: string | null;
  incidente_folder: string | null;
  incidente_cnj: string | null;
  incidente_id: number | null;
  confirmado_candidato_id: number | null;
  anotacao: string | null;
  criado_em: string | null;
  candidatos?: number;
  candidatos_fortes_pendentes?: number;
  partes_demandadas?: number;
}

export interface EmbargosParametros {
  corte_relatorio: string | null;
  janela_dias_uteis: number;
  janelas_permitidas: number[];
  intervalo_dias_uteis: number;
  teto_dias: number;
  aviso_emails: string[];
  relatorio_titulo: string;
  relatorio_modelo_id: number;
}

export interface RelatorioUltimo {
  ok?: boolean;
  report_id?: number;
  novas?: number;
  atualizadas?: number;
  antes_do_corte?: number;
  sem_dados?: boolean;
  erro?: string;
  em?: string;
  corte?: string | null;
}

export interface RelatorioStatus {
  running?: boolean;
  fase?: string | null;
  iniciado_em?: string | null;
  ultimo?: RelatorioUltimo | null;
}

export interface PartesStatus {
  running?: boolean;
  iniciado_em?: string | null;
  na_fila?: number;
  origem?: string;
  ultimo?: { na_fila?: number; origem?: string; em?: string; erro?: string } | null;
}

export interface EmbargosListaResponse {
  total: number;
  kpis: {
    por_estado: Record<string, number>;
    partes_erro: number;
    consultas_vencidas: number;
    total: number;
  };
  parametros: EmbargosParametros;
  relatorio: RelatorioStatus;
  partes: PartesStatus;
  items: EmbargosItem[];
}

export interface EmbargosParte {
  id: number;
  origem: string;
  polo: string | null;
  nome: string;
  cpf_cnpj: string | null;
  tipo_pessoa: string | null;
  relacao_bb: string | null;
  demandada: boolean;
}

export interface EmbargosCandidato {
  id: number;
  cnj: string;
  data_ajuizamento: string | null;
  classe_nome: string | null;
  orgao_nome: string | null;
  distribuicao_dependencia: boolean;
  peticao_mesmo_dia: boolean;
  nivel: string;
  djen_status: string | null;
  djen_embargantes: string[];
  djen_embargados: string[];
  l1_litigation_id: number | null;
  l1_folder: string | null;
  djen_nomes_casados: string[];
  djen_trecho: string | null;
  djen_data: string | null;
  djen_consultado_em: string | null;
  decisao: string;
  decidido_em: string | null;
  criado_em: string | null;
}

export interface EmbargosEvento {
  id: number;
  secao: string;
  nivel: string;
  mensagem: string;
  dados: Record<string, unknown> | null;
  user_id: number | null;
  criado_em: string | null;
}

export interface EmbargosDisparo {
  id: number;
  template_nome: string | null;
  subtipo_id: number | null;
  incidente_id: number | null;
  responsavel_contact_id: number | null;
  prazo: string | null;
  descricao: string | null;
  status: string; // CRIADA | FALHA
  l1_task_id: number | null;
  erro: string | null;
  criado_em: string | null;
}

export interface EmbargosDetalhe {
  execucao: EmbargosItem;
  l1_task_ids: number[];
  disparos: EmbargosDisparo[];
  partes: EmbargosParte[];
  candidatos: EmbargosCandidato[];
  eventos: EmbargosEvento[];
  partes_disparada?: boolean;
  resultado_disparo?: { criadas: number; falhas: number; puladas: number; estado: string };
}

export interface EmbargosTemplate {
  id: number;
  nome: string;
  ativo: boolean;
  ordem: number;
  tipo_id: number;
  subtipo_id: number;
  subtipo_nome: string | null;
  responsavel_modo: string; // FIXO | ADVOGADO_RESPONSAVEL
  responsavel_contact_id: number | null;
  responsavel_nome: string | null;
  prazo_dias_uteis: number;
  prioridade: string;
  descricao_template: string;
  observacoes_template: string | null;
  atualizado_em: string | null;
}

export interface EmbargosTemplatesResponse {
  total: number;
  items: EmbargosTemplate[];
  placeholders: Record<string, string>;
  prioridades: string[];
  responsavel_modos: Record<string, string>;
}

export interface EmbargosTemplateIn {
  nome: string;
  ativo: boolean;
  ordem: number;
  subtipo_id: number;
  responsavel_modo: string;
  responsavel_contact_id: number | null;
  prazo_dias_uteis: number;
  prioridade: string;
  descricao_template: string;
  observacoes_template: string | null;
}

export interface EmbargosPreviaTarefa {
  template_id: number;
  nome: string;
  subtipo_nome: string | null;
  responsavel_contact_id: number | null;
  responsavel_nome: string | null;
  prazo: string;
  prioridade: string;
  descricao: string;
  observacoes: string | null;
  erro: string | null;
  ja_disparada: boolean;
  l1_task_id: number | null;
}

export interface EmbargosPrevia {
  incidente: { id: number; folder: string | null; cnj: string | null; office_id: number | null } | null;
  tarefas: EmbargosPreviaTarefa[];
  placeholders: Record<string, string>;
}

export interface EmbargosListaParams {
  estado?: string;
  cliente?: string;
  partes_status?: string;
  busca?: string;
  ordenar?: string;
  direcao?: "asc" | "desc";
  limit?: number;
  offset?: number;
}

export async function listarEmbargos(params: EmbargosListaParams): Promise<EmbargosListaResponse> {
  const qs = new URLSearchParams();
  if (params.estado) qs.set("estado", params.estado);
  if (params.cliente) qs.set("cliente", params.cliente);
  if (params.partes_status) qs.set("partes_status", params.partes_status);
  if (params.busca) qs.set("busca", params.busca);
  if (params.ordenar) qs.set("ordenar", params.ordenar);
  if (params.direcao) qs.set("direcao", params.direcao);
  qs.set("limit", String(params.limit ?? 50));
  qs.set("offset", String(params.offset ?? 0));
  return json(await apiFetch(`${BASE}?${qs.toString()}`));
}

export async function getEmbargosDetalhe(id: number): Promise<EmbargosDetalhe> {
  return json(await apiFetch(`${BASE}/${id}`));
}

export async function getEmbargosParametros(): Promise<EmbargosParametros> {
  return json(await apiFetch(`${BASE}/parametros`));
}

export async function salvarEmbargosParametros(
  body: Partial<Pick<EmbargosParametros, "corte_relatorio" | "janela_dias_uteis" | "intervalo_dias_uteis" | "teto_dias" | "aviso_emails">>,
): Promise<EmbargosParametros> {
  return json(await apiFetch(`${BASE}/parametros`, { method: "PUT", body: JSON.stringify(body) }));
}

export async function getRelatorioStatus(): Promise<RelatorioStatus> {
  return json(await apiFetch(`${BASE}/relatorio/status`));
}

export async function gerarRelatorioEmbargos(): Promise<{ ok: boolean; mensagem: string }> {
  return json(await apiFetch(`${BASE}/relatorio/gerar`, { method: "POST" }));
}

export async function getPartesStatus(): Promise<PartesStatus> {
  return json(await apiFetch(`${BASE}/partes/status`));
}

export async function coletarPartesAgora(): Promise<{ ok: boolean; mensagem: string }> {
  return json(await apiFetch(`${BASE}/partes/coletar`, { method: "POST" }));
}

export async function importarPlanilhaEmbargos(file: File): Promise<Record<string, unknown> & { novas: number; atualizadas: number; sem_data: number; sem_identificador: number }> {
  const fd = new FormData();
  fd.append("arquivo", file);
  return json(await apiFetch(`${BASE}/importar`, { method: "POST", body: fd }));
}

export async function incluirEmbargosManual(body: {
  pasta?: string;
  cnj?: string;
  npj?: string;
  data_ajuizamento: string;
}): Promise<{ novas: number; atualizadas: number }> {
  return json(await apiFetch(`${BASE}/manual`, { method: "POST", body: JSON.stringify(body) }));
}

export async function ajustarEmbargos(
  id: number,
  body: { dias_uteis_janela?: number; anotacao?: string },
): Promise<EmbargosDetalhe> {
  return json(await apiFetch(`${BASE}/${id}`, { method: "PATCH", body: JSON.stringify(body) }));
}

export async function consultarEmbargosAgora(id: number): Promise<{ ok: boolean; mensagem: string }> {
  return json(await apiFetch(`${BASE}/${id}/consultar`, { method: "POST" }));
}

export async function recoletarPartesEmbargos(id: number): Promise<EmbargosDetalhe> {
  return json(await apiFetch(`${BASE}/${id}/partes`, { method: "POST" }));
}

export async function decidirCandidatoEmbargos(
  id: number,
  candidatoId: number,
  decisao: "CONFIRMADO" | "RECUSADO",
): Promise<EmbargosDetalhe> {
  return json(
    await apiFetch(`${BASE}/${id}/candidatos/${candidatoId}/decisao`, {
      method: "POST",
      body: JSON.stringify({ decisao }),
    }),
  );
}

export async function encerrarEmbargos(id: number, motivo?: string): Promise<EmbargosDetalhe> {
  return json(await apiFetch(`${BASE}/${id}/encerrar`, { method: "POST", body: JSON.stringify({ motivo }) }));
}

export async function reabrirEmbargos(id: number): Promise<EmbargosDetalhe> {
  return json(await apiFetch(`${BASE}/${id}/reabrir`, { method: "POST" }));
}

export async function previaTarefasEmbargos(id: number): Promise<EmbargosPrevia> {
  return json(await apiFetch(`${BASE}/${id}/tarefas/previa`));
}

export async function dispararTarefasEmbargos(id: number): Promise<EmbargosDetalhe> {
  return json(await apiFetch(`${BASE}/${id}/tarefas/disparar`, { method: "POST" }));
}

export async function listarTemplatesEmbargos(): Promise<EmbargosTemplatesResponse> {
  return json(await apiFetch(`${BASE}/templates`));
}

export async function salvarTemplateEmbargos(body: EmbargosTemplateIn, id?: number): Promise<EmbargosTemplate> {
  return json(
    await apiFetch(id ? `${BASE}/templates/${id}` : `${BASE}/templates`, {
      method: id ? "PUT" : "POST",
      body: JSON.stringify(body),
    }),
  );
}

export async function excluirTemplateEmbargos(id: number): Promise<{ ok: boolean }> {
  return json(await apiFetch(`${BASE}/templates/${id}`, { method: "DELETE" }));
}

// Catálogos compartilhados com os templates de Publicações (qualquer usuário logado).
export interface CatalogoTipo {
  external_id: number;
  name: string;
  subtypes: { external_id: number; name: string }[];
}
export interface CatalogoUsuario {
  id: number;
  external_id: number | null;
  name: string;
  email?: string | null;
  squads?: { id: number; name: string }[];
}

export async function getTiposTarefa(): Promise<CatalogoTipo[]> {
  return json(await apiFetch(`/api/v1/task-templates/meta/task-types`));
}

export async function getUsuariosL1(): Promise<CatalogoUsuario[]> {
  return json(await apiFetch(`/api/v1/task-templates/meta/users`));
}
